### PR TITLE
[MM-38103] initialize select fields with the value

### PR DESCRIPTION
**Pull request by jfrerich**
#### Summary
This PR fixes a bug where select modal values were not getting set

The change can be tested with https://github.com/mattermost/mattermost-app-test/pull/12

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-38102

#### Device Information
This PR was tested on: Android Emulator

#### Screenshots
Before:
 

After: 
 

#### Release Note
```release-note
Fixed a bug where the select modal value was not getting populated
```

### CODE REVIEW

File: app/screens/apps_form/apps_form_field.tsx [hunk 1]
```diff
@@ -49,9 +49,28 @@ const getStyleSheet = makeStyleSheetFromTheme((theme: Theme) => {
...
+            const value = props.value as AppSelectOption | null;
+            if (value) {
+                selected = {
+                    text: value.label,
+                    value: value.value,
+                };
+            }
+        }
+        }
+        this.state = {
```
Comment (mickmister) [inline app/screens/apps_form/apps_form_field.tsx:70]: 0/5 this may look better with a newline here

```suggestion

        this.state = {
```

**Review by mickmister — APPROVED** [review]
LGTM, just one code styling suggestion

File: app/screens/apps_form/apps_form_field.tsx (hunk 1 above)
Comment (jfrerich) [reply app/screens/apps_form/apps_form_field.tsx:71]: done.

**Review by larkox — APPROVED** [review]
LGTM. I don't know why I thought this was already solved somewhere else.

**Review by mickmister — COMMENTED** [review]
@larkox @jfrerich I think this PR needs to support multiselect values as well

**Comment by larkox** [comment]
Multiselect for mobile is not yet merged (https://github.com/mattermost/mattermost-mobile/pull/5400).

**Comment by mickmister** [comment]
> Multiselect for mobile is not yet merged (#5400).

It is now 🙂

**Comment by jfrerich** [comment]
@mickmister, @larkox I have added the initialization of multiselect also.  Looking at `<AutocompleteSelector>`, the `selected` prop expects an array if multiselect and one object if not multiselect. I accounted for this in `app_form_field.tsx`  I wonder if it would have been better to always pass `DialogOption[]` to the `selected` prop instead?  If so, maybe we could create a HW ticket to make that change?

**Comment by mickmister** [comment]
> I wonder if it would have been better to always pass `DialogOption[]` to the `selected` prop instead? If so, maybe we could create a HW ticket to make that change?

We might as well add it to this PR since it's still open. Can you propose the change you're referring to via the PR review feature?

File: app/components/autocomplete_selector/autocomplete_selector.tsx [hunk 2]
```diff
@@ -47,7 +47,7 @@ type Props = {
 
 type State = {
     selectedText: string;
-    selected?: DialogOption | DialogOption[];
+    selected?: DialogOption[];
```
Comment (jfrerich) [inline app/components/autocomplete_selector/autocomplete_selector.tsx:50]: Always accept an array of DialogOtions.

File: app/components/autocomplete_selector/autocomplete_selector.tsx [hunk 3]
```diff
@@ -74,14 +74,17 @@ export default class AutocompleteSelector extends PureComponent<Props, State> {
             return null;
         }
 
+        const options = props.selected as DialogOption[];
         if (!props.isMultiselect) {
-            return {
-                selectedText: (props.selected as DialogOption).text,
-                selected: props.selected,
-            };
+            if (options.length === 1) {
```
Comment (jfrerich) [inline app/components/autocomplete_selector/autocomplete_selector.tsx:79]: field is not multiselect. if one value use it, else value is null.

File: app/screens/apps_form/apps_form_field.tsx [hunk 4]
```diff
@@ -49,49 +49,39 @@ const getStyleSheet = makeStyleSheetFromTheme((theme: Theme) => {
...
+            selected,
         };
+    }
 
-        this.props.onChange(field.name, selectedOption);
-    };
-
-    handleMultioptionAutocompleteSelect = (selected: DialogOption[]) => {
+    handleAutocompleteSelect = (selected: DialogOption[]) => {
```
Comment (jfrerich) [inline app/screens/apps_form/apps_form_field.tsx:78]: handleAutocompleteSelect is now simple.  It only changes sets the select value which is passed to `AutocompleteSelect`

File: app/screens/apps_form/apps_form_field.tsx [hunk 5]
```diff
@@ -210,7 +200,7 @@ export default class AppsFormField extends React.PureComponent<Props, State> {
                     dataSource={dataSource}
                     options={options}
                     optional={!field.is_required}
-                    onSelected={field.multiselect ? this.handleMultioptionAutocompleteSelect : this.handleAutocompleteSelect}
+                    onSelected={this.handleAutocompleteSelect}
```
Comment (jfrerich) [inline app/screens/apps_form/apps_form_field.tsx:203]: `this.handleMultioptionAutocompleteSelect` now one function.  `AutocompleteSelect` will handle if value is multiselect

**Comment by jfrerich** [comment]
@mickmister, @larkox I left comments on the latest commit.  If y'all don't like this pattern, I can easily revert.

File: app/components/autocomplete_selector/autocomplete_selector.tsx (hunk 2 above)
Comment (larkox) [reply app/components/autocomplete_selector/autocomplete_selector.tsx:50]: This makes things simpler on our code, but does it make things simpler for the app developer? I mean... the general case will be something like `selected: [{label: "someLabel", value:"someValue"}]`. Is the value this change adds worth the extra `[]` on the developer experience.

Comment (jfrerich) [reply app/components/autocomplete_selector/autocomplete_selector.tsx]: After thinking about it, I agree it makes the code simpler, but having the developer add stuff one object in an array isn't worth it.  

I reverted this change. Please re-review again. I added the multiselect handling.

File: app/screens/apps_form/apps_form_field.tsx [hunk 6]
```diff
@@ -49,10 +49,33 @@ const getStyleSheet = makeStyleSheetFromTheme((theme: Theme) => {
...
         super(props);
 
-        this.state = {};
+        let selected;
+        switch (props.field.type) {
+        case AppFieldTypes.STATIC_SELECT:
+        case AppFieldTypes.DYNAMIC_SELECT:
+        case AppFieldTypes.USER:
+        case AppFieldTypes.CHANNEL: {
+            const value = props.value as AppSelectOption[] | null;
```
Comment (mickmister) [inline app/screens/apps_form/apps_form_field.tsx:61]: So there's no case where `value` is a single object for when it's not a multiselect? What is doing the conversion from a single object to an array? It looks like from the thread above it was decided that the App dev should not have to use an array in the single value case.

Comment (jfrerich) [reply app/screens/apps_form/apps_form_field.tsx]: Corrected in the latest commit.

**Review by larkox — APPROVED** [review]

**Comment by enahum** [comment]
@DHaussermann friendly reminder to review this PR

**Comment by hanzei** [comment]
@DHaussermann Gentle reminder to review this PR

**Comment by enahum** [comment]
@DHaussermann is this still on your radar?

**Comment by larkox** [comment]
@mickmister With V2 in beta, I wonder if this work is still needed for master, and if we should move it for the effort that has to be made directly in V2. Thoughts?

**Comment by enahum** [comment]
@DHaussermann did you try it?

**Comment by DHaussermann** [comment]
@enahum Yes I did try it. The build showed no issue in Ci so I did not ask dev for help but then the links never posted in the Build channel?

I'm trying this again now. If I don't see the links post I'll try to just go directlky yo Ci

**Review by DHaussermann — APPROVED** [review]
Tested and passed
- Muti-select values can be set - Confirmed in UI and test tooling shows .json payload in a request
- Ensured no issues editing/changing the value selection of a mutli-select before form submission
- No issues discarding edit with back navigation
- Regression tested single select control such as used fir a single dynamic select value - No issues found
- Tested on Android and iOS 
LGTM!

@larkox and @enahum I'm sorry for the delay in bumping this higher in my queue and getting it tested.  
Please merge.

**Pull request merged**